packages/devtools_app/lib/src/charts/chart_trace.dart

```diff
@@ -0,0 +1,213 @@
+
+  /// If specified Y scale is computed and max value is fixed.
+  /// Will assert if new data point is more than max specified.
+  double fixedMaxY;
+
+  Color color;
+
+  ChartSymbol symbol;
+
+  double strokeWidth;
+
+  double diameter;
+}
+
+class Trace {
+  Trace(this.controller, this._chartType, this.characteristics) {
+    double minY = characteristics.fixedMinY;
+    minY ??= 0.0;
+
+    double maxY = characteristics.fixedMaxY;
+    maxY ??= 0.0;
+
+    yAxis = AxisScale(minY, maxY, 30);
+  }
+
+  final ChartController controller;
+
+  final ChartType _chartType;
+
+  final PaintCharacteristics characteristics;
+
+  String name;
+
+  double dataYMax = 0;
+
+  final data = <Data>[];
+
+  ChartType get chartType => _chartType;
+
+  AxisScale yAxis;
+
+  void addDatum(Data datum) {
+    data.add(datum);
+
+    if (characteristics.fixedMaxY != null) {
+      assert(
+        datum.y <= characteristics.fixedMaxY,
+        'y=${datum.y} fixedMaxY=${characteristics.fixedMaxY}',
+      );
+    } else if (datum.y > dataYMax) {
+      dataYMax = datum.y.toDouble();
+      yAxis = AxisScale(0, dataYMax, 30);
+    }
+
+    if (datum.y > controller?.yMaxValue) controller?.yMaxValue = datum.y;
+
+    final traceIndex = controller.traceIndex(this);
+
+    // New data has arrived notify listeners this data needs to be plotted.
+    controller.traceChanged.value = TraceNotifier(traceIndex, data.length - 1);
+  }
+}
+
+class TraceNotifier {
+  TraceNotifier(this.traceIndex, this.dataIndex);
+
+  int traceIndex;
+  int dataIndex;
+}
+
+enum ChartType {
+  symbol,
+  line,
+}
+
+enum ChartSymbol {
+  ring,
+  disc,
+  square,
+  triangle,
+  dashedLine,
+}
+
+double log10(num x) => log(x) / ln10;
+
+class AxisScale {
+  AxisScale(this.minPoint, this.maxPoint, this.maxTicks) {
+    _calculate();
+  }
+
+  final double minPoint, maxPoint;
+
+  final double maxTicks;
+
+  double _computedMin, _computedMax;
+
+  double _tickSpacing;
+
+  double get tickSpacing => _tickSpacing;
+  double get computedMin => _computedMin;
+  double get computedMax => _computedMax;
+
+  double _range;
+
+  /// Unit for the label (exponent) e.g., 6 = 10^6
+  double labelUnitExponent;
+
+  /// Number of lables.
+  double labelTicks;
+
+  void _calculate() {
+    _range = _niceNum(maxPoint, false);
+    if (_range != 0) {
+      _tickSpacing = _range / (maxTicks - 1);
+      _computedMin = _calculateMin();
+      _computedMax = _calculateMax();
+      final exponentFraction = _exponentFraction(_range);
+      labelUnitExponent = exponentFraction['exponent'];
+      labelTicks = exponentFraction['fraction'].roundToDouble();
+    } else {
+      _tickSpacing = 1;
+      _computedMin = 0;
+      _computedMax = 0;
+      labelUnitExponent = 0;
+      labelTicks = 0;
+    }
+  }
+
+  double _calculateMin() {
+    final floored = (minPoint / _tickSpacing).floor();
+    return floored * _tickSpacing;
+  }
+
+  double _calculateMax() {
+    final ceiled = (maxPoint / _tickSpacing).ceil();
```

Comment:
nit remove variable ceiled that doesn't clarify anything. The question I have is why the max is intentionally past the maxPoint. A quick comment indicating why that is intentional could be helpful.

Reply:
Done.